rlog/2023-09-27-gossipimprovements.mdx

```diff
@@ -0,0 +1,214 @@
+
+The properties and nature of the overlay network topology significantly impact the performance of services and applications executed on top of them. 
+Studying and devising mechanisms for better overlay design and message dissemination is paramount to achieving superior performance.
+
+Interestingly, shortest-path message delivery trees have many limitations: 
+
+1) Changing network dynamics requires a quicker and continuous readjustment of the multicast tree. 
+2) The presence of resource-constrained (bandwidth/compute, etc.) nodes in the overlay can result in congestion. 
+3) Node failure can result in partitions, making many segments unreachable. 
+4) Assuring a shortest-path tree-like structure requires a detailed view of the underlying (and continuously changing) network topology. 
+
+
+Solutions involve creating multiple random trees to add redundancy [2].
+Alternatives involve building an overlay mesh and forwarding messages through the multicast delivery tree (eager push). 
+
+Metadata is shared through the overlay links so that the nodes can ask for missing messages (lazy push or pull-based operation) through the overlay links. 
+New nodes are added from the overlay on node failure, but it requires non-faulty node selection.
+
+GossipSub uses eager push (through overlay mesh) and lazy push (through IWANT messages). 
+
+The mesh degree $D_{Low} \leq D \leq D_{High}$ is crucial in deciding message dissemination latency. 
+A smaller value for $D$ results in higher latency due to increased rounds, whereas a higher $D$ reduces latency on the cost of increased redundancy. 
+It is suggested that the average mesh degree should be $D_{avg} = \ln(N) + C$ for an optimal operation, where $N$ is the network size and C is a small constant. 
+
+Moreover, quicker shuffling of peers results in better performance in the presence of resource-constrained nodes and/or node failure [3]. 
+
+## GOAL2: Considering Heterogeneity In Overlay Design
+Random peering connections in P2P overlays represent a stochastic process. It is inherently difficult to precisely model the performance of such systems. 
+Most of the research on P2P networks provides simulation results assuming nodes with similar capabilities. 
+The aspect of dissimilar capabilities and resource-constrained nodes is less explored.
+
+It is discussed in GOAL1 that overlay mesh results in better performance if $D_{avg} = \ln(N) + C$. 
+Enforcing all the nodes to have approximately $\ln(N) + C$ peers makes resource-rich nodes under-utilized, while resource-constrained nodes are overloaded. 
+At the same time, connecting high-bandwidth nodes through a low-bandwidth node undermines the network's performance.
+Ideally, every node should contribute proportionally to the available resources.
+A better solution involves a two-phased operation:
+
+
+1. Every node computes its available bandwidth and selects a node degree $D_i$ proportional to its available bandwidth [3].

```

Comment:
That would only be applicable in networks where participants have "bandwidth to spare"

Generally, home users don't want to spend all of their bandwidth to participate in some network (though maybe we could look at something like µTP)

And datacenter users tend to pay for their bandwidth, so even though they have a lot at hand, they might not want to "waste it"


However, I can see a benefit to lowering D locally if the node doesn't has enough bandwidth, as this would slow down the network either way

Reply:
ba67fe604acb1211a2617f461adc08b8097f1e47